### ISSUE TITLE
ICOFamily created with bitmap image reps are corrupt

### ISSUE THREAD

**Issue by danwood** [alexzielenski/ICOFamily#4]
There is something wrong with creating an ICOFamily out of specific image sizes rather than the test code which just creates a family from a single NSImage.

I have set up a branch on my fork of ICOFamily called "colorfix" that demonstrates this problem.  I'm creating an empty ICOFamily, and then trying to add in 16 pixel and 32 pixel elements from bitmaps.

https://github.com/karelia/ICOFamily/tree/colorfix

First off, there is a problem with the call to -[NSBitmapImageRep bitmapImageRepByConvertingToColorSpace:renderingIntent:] where you get this logged:

&lt;Error&gt;: CGBitmapContextCreate: invalid data bytes/row: should be at least 128 for 16 integer bits/component, 3 components, kCGImageAlphaPremultipliedLast.

So this is kind of mysterious.  If I comment out this line, since it's sort of optional, I get a file written out, but it's corrupt and Preview can't render it.

I'll plug away at this a bit, but if @alexzielenski or @matheweis have any thoughts, I could use some help. :-)

**Comment by matheweis** [comment]
Some quick thoughts:

1) You're using 16-bit/channel png images, so...
2) .... looks like a Cocoa bug in bitmapImageRepByConvertingToColorSpace:renderingIntent - it can't handle the 16-bit/channel conversion right
3) ... neither can my (admittedly very basic) code for - (NSData*)data (It assumes 8-bit/channel RGBA)

If you use 8-bit/channel images, it should work better.

**Comment by matheweis** [comment]
The best short term solution would probably be to force it to bring in the NSBitmapImageRep as 8-bit instead of XX-bit - very much as ICOFamily already does in
- (void)setImage:(NSImage*)image forElement:(kICOFamilyElement)element

instead of using [[[NSBitmapImageRep alloc] initWithCGImage:rep.CGImage] inside of:
- (void)setBitmapImageRep:(NSBitmapImageRep*)rep forElement:(kICOFamilyElement)element

**Comment by danwood** [comment]
Yeah, from my test which I just checked into the karelia/colorfix branch, setImage:forElement: seems to do the trick.  Though next I will test the output with browsershots.org to make sure that they _really_ work as favicons :-)
